### PR TITLE
W.I.P. hdf5 backend code generation

### PR DESCRIPTION
**Pull request by hodger**
PR for collaboration. Not yet functional or ready for merging. (@scopatz)

This will allow for more effective hdf5 backend generation by using an AST setup with various nodes and blocks of nodes. The recursive generation of a type depends on the generation of a setup, declaration, body, and teardown, which build the type based on its sub types (and their sub types, and so on).

### CODE REVIEW

File: src/hdf5_back_gen.py [hunk 1]
```diff
@@ -3,6 +3,279 @@
...
+class DeclAssign(Node):
+    fields = ("type", "target", "value")
+
+class Case(Node):
+    fields = ("cond", "body")
+    
+class If(Node):
+    fields = ("cond", "body", "elifs", "el")
+
+#how to allow ';' as decl
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: PEP8 single space after comment.

**Comment by scopatz** [comment]
Thanks for putting this in @hodger!

File: src/hdf5_back_gen.py [hunk 2]
```diff
@@ -3,6 +3,279 @@
...
+        for n in node.body:
+            s += indent(self.visit(n), self.indent)
+        s += "\n}\n"
+        return s
+        
+    def visit_funccall(self, node):
+        s = self.visit(node.name)
+        if node.targs is not None:
+            s += "<"
+            for i in range(0, len(node.targs)):
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: No need to include the `0` in the range, it is implicit.

File: src/hdf5_back_gen.py [hunk 3]
```diff
@@ -3,6 +3,279 @@
...
+            s += ">"
+        s += "("
+        for i in range(0, len(node.args)):
+            s += self.visit(node.args[i])
+            if i < (len(node.args)-1):
+                s += ","
+        s += ")"
+        return s
+    
+    def visit_nothing(self, node):
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Sounds like a staycation :)

File: src/hdf5_back_gen.py [hunk 4]
```diff
@@ -106,405 +386,338 @@ def __init__(self, canon):
...
-hsize_t fieldlen;
-H5Tget_array_dims2(field_type, &fieldlen);
-unsigned int itemsize = 2*CYCLUS_SHA1_SIZE;
-{t.cpp} x;
-for(unsigned int k = 0; k < fieldlen; ++k) {{
-    x[VLRead<{t.sub[1].cpp}, {t.sub[1].db}>(buf + offset + itemsize*k)] = VLRead<{t.sub[2].cpp}, {t.sub[2].db}>(buf + offset + itemsize*k + CYCLUS_SHA1_SIZE);
-}}
-{H5TCLOSE}
+def normal_close(t):
+    """
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Docstrings could be a little more descriptive, such as "This function represents..."

File: src/hdf5_back_gen.py [hunk 5]
```diff
@@ -106,405 +386,338 @@ def __init__(self, canon):
...
+    return tree
+
+def vector_reader(t):
+    """
+    {t.cpp} x = {t.cpp}(col_sizes_[table][j] / sizeof({t.sub[1].cpp}));
+    memcpy(&x[0], buf + offset, col_sizes_[table][j]);
+    {NO_CLOSE}
+    """
+    tree = Block(nodes=[
+                 #setup
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: I don't think that these comments are needed anymore.

File: src/hdf5_back_gen.py [hunk 6]
```diff
@@ -106,405 +386,338 @@ def __init__(self, canon):
...
+#           'UUID': UUID_READER,
+#           'VECTOR': VECTOR_READER,
+#           'VL_VECTOR': VL_READER,
+#           'SET': SET_READER,
+#           'VL_SET': VL_READER,
+#           'LIST': LIST_READER,
+#           'VL_LIST': VL_READER,
+#           'PAIR': PAIR_READER,
+#           'MAP': MAP_READER,
+#           'VL_MAP': VL_READER}
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Please remove

**Comment by scopatz** [comment]
Overall, it looks very good other than the failures, @hodger :)

**Comment by hodger** [comment]
@scopatz Thanks for looking over this. Got most of the PEP8 stuff. I will update the function comments tonight. Sorry it has taken me a while to get going on this again- finally getting situated with everything here in ND.

**Comment by scopatz** [comment]
No worries @hodger!  Thanks for the effort here. Just ping when you think it is ready for another look.

**Comment by hodger** [comment]
@scopatz Started jumping back into this over the past couple weeks! 

Working on a `get_size` function to help provide nested types with the correct offset to use for initialization. For types which involve a loop, I noticed that they loop either `jlen` or `fieldlen` times. Fieldlen seems like it is used when a normal std::string is involved, while jlen is used otherwise (except in the case of MAP_VL_STRING_VL_STRING, where I noticed fieldlen is used). Is this a correct observation to make?

My current idea is to potentially create a `jlen`/`fieldlen` variable for every type during setup (most likely recursively), as well as an `itemsize` variable for each type using that value. Then, a higher level type will know which offset to use for its children by simply calling itemsize with the proper variable prefix.

As an example, you might have PAIR_VECTOR_VECTOR_INT_DOUBLE.

VECTOR_INT has a `jlen1` value and its `itemsize` = `size-of(int) * jlen1`. 

Currently, `jlen1` is a variable created during the setup of the int rather than the vector itself. That might not be the best place for it.

To know how to initialize DOUBLE, we need to know the size of VECTOR_VECTOR_INT to use as the offset. The itemsize of VECTOR_VECTOR_INT needs to be `(size-of(int) * jlen1) * jlen0`. 

We don't currently have `jlen0`, but it would be available if we create it during the setup step for the vector.

**Comment by hodger** [comment]
@scopatz All types generate and compile now, but the hdf5_back_tests are throwing HDF5 errors. These seem to mostly concern the use of functions like `H5Tget_member_type`.

For instance, the generated VECTOR_INT code is this:
```
case VECTOR_INT: {
  hid_t fieldtype0=H5Tget_member_type(tb_type,j);
  unsigned int total_size0=H5Tget_size(fieldtype0);
  hsize_t fieldlen0;
  H5Tget_array_dims2(fieldtype0,&fieldlen0);
  unsigned int item_size0=total_size0/fieldlen0;
  hid_t fieldtype1ELEM=H5Tget_member_type(fieldtype0,0);
  unsigned int total_size1ELEM=H5Tget_size(fieldtype1ELEM);
  std::vector<int> x0;
  x0=std::vector<int>(fieldlen0);
  memcpy(&x0[0],buf+offset,total_size0);
  is_row_selected=CmpConds<std::vector<int>>(&x0,&(field_conds[qr.fields[j]]));
  if(is_row_selected){
    row[j]=x0;

  }
  H5Tclose(fieldtype1ELEM);
  H5Tclose(fieldtype0);
  break;

}
```
And it gives the following errors:
```
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5Tcompound.c line 248 in H5Tget_member_type(): not a compound datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5T.c line 2142 in H5Tget_size(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5T.c line 1761 in H5Tclose(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5Tcompound.c line 248 in H5Tget_member_type(): not a compound datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5T.c line 2142 in H5Tget_size(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.11) thread 139699674433408:
  #000: ../../../src/H5T.c line 1761 in H5Tclose(): not a datatype
    major: Invalid arguments to routine
    minor: Inappropriate type
```
Is something like VECTOR_INT not considered a compound data type? I've been looking at documentation but I haven't found anything helpful yet.

**Comment by hodger** [comment]
@scopatz I cleaned up the formatting, added some docstrings, and made the body function a little more recursive. Can you take a look?

File: src/hdf5_back_gen.py [hunk 7]
```diff
@@ -78,433 +354,735 @@ def list_dependencies(canon):
...
+
+def get_prefix(base_prefix, parent_type, child_index):
+    """Return the prefix of a C++ variable, appropriately formatted for depth.
+    """
+    return base_prefix + template_args[parent_type.canon[0]][child_index]
+
+def case_template(t, read_x):
+    """Represents C++ case statement
+    
+    Parameters:
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: This should be in numpydoc format.

File: src/hdf5_back_gen.py [hunk 8]
```diff
@@ -78,433 +354,735 @@ def list_dependencies(canon):
...
+variable_length_types = ["MAP", "LIST", "SET", "VECTOR"]
+
+def get_setup(t, depth=0, prefix="", HDF5_type="tb_type", child_index='j'):
+    """Get nodes representing C++ setup. 
+    
+    Primitive setups are called directly, while template types are handled
+    recursively.
     
     Parameters
-    ----------
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: This should be in numpydoc format

File: src/hdf5_back_gen.py [hunk 9]
```diff
@@ -78,433 +354,735 @@ def list_dependencies(canon):
...
+        
+        node = Block(nodes=setup_nodes)
+    return node
+
+def get_decl(t, depth=0, prefix=""):
+    """Get node representing C++ type declaration. 
+    
+    Declarations occur directly before bodies, created without recursion.
+    
+    Parameters:
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: numpydoc

**Comment by scopatz** [comment]
Other than some of the docstring formatting issues, this looks great to me!

**Comment by hodger** [comment]
Gotcha, I'll correct the formatting

**Comment by hodger** [comment]
@scopatz Adjusted those docstrings. I didn't give every body and setup function a full docstring, since their parameters and returns are the functionally the same as their base `get_body` and `get_setup` functions.

**Comment by scopatz** [comment]
Thanks so much @hodger! Really great work here!

**Comment by hodger** [comment]
Thanks!

**Pull request merged**